### PR TITLE
feat(cli): support hotswapping Lambda function tags

### LINKED ISSUE #17664: (cli): Hotswap support for resource tags
### Description

I would like hotswap to modify resource tags in-place rather than triggering a full stack changeset. Assuming a diff like this:

```
cdk diff mystack
Including dependency stacks: mystack-Data
Stack mystack-Data
There were no differences
Stack mystack
Resources
[~] AWS::Lambda::Function MyApp MyAppE55F03CE
 ├─ [~] Code
 │   └─ [~] .S3Key:
 │       ├─ [-] local_development_resourcesd5d9014617720b099ab57419bf3bfb4beb0db64d0290ea1693040723b4489f1c
 │       └─ [+] local_development_resources008d8f94dad3437e716aef4e7280656cee4106115ad42f6aafc94b9e14108397
 └─ [~] Tags
     └─ @@ -5,6 +5,6 @@
        [ ]   },
        [ ]   {
        [ ]     "Key": "tagName",
        [-]     "Value": "local_development_resourcesd5d9014617720b099ab57419bf3bfb4beb0db64d0290ea1693040723b4489f1c"
        [+]     "Value": "local_development_resources008d8f94dad3437e716aef4e7280656cee4106115ad42f6aafc94b9e14108397"
        [ ]   }
        [ ] ]
```

This delta should be safe to hotswap, but because there is a tag change, it does not.

### Use Case

My stack assets infer tag values from asset hashes, and I would still like to be able to hotswap without having to have a separate code path for hotswapped assets.

### Proposed Solution

Include the `Tags` in values that are hotswappable, across all resource types.

### Other information

_No response_

### Acknowledge

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by offbyone**
Fixes #17664 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by offbyone** [comment]
This will need at least one more revision before it's ready; hold off on review til EOD today, at least, please.

**Comment by skinny85** [comment]
> This will need at least one more revision before it's ready; hold off on review til EOD today, at least, please.

Sound good @offbyone, let us know when you consider this ready!

**Comment by offbyone** [comment]
I'd call it ready for review. Despite our discussion about the asset tags in the metadata, I was unable to avoid removing that check. I believe the property checks in here are sufficient safety.

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts [hunk 1]
```diff
@@ -64,43 +69,94 @@ async function isLambdaFunctionCodeOnlyChange(
...
-          s3Bucket = await evaluateCfnTemplate.evaluateCfnExpression(updatedProp.newValue[newPropName]);
-          break;
-        case 'S3Key':
-          foundCodeDifference = true;
-          s3Key = await evaluateCfnTemplate.evaluateCfnExpression(updatedProp.newValue[newPropName]);
-          break;
-        default:
-          return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
-      }
+
```
Comment (skinny85) [inline packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:78]: Again, no need for this empty line.

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts [hunk 2]
```diff
@@ -64,43 +69,94 @@ async function isLambdaFunctionCodeOnlyChange(
...
+              return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+          }
+        }
+        break;
+      case 'Tags':
+        /*
+         * Tag updates are a bit odd; they manifest as two lists, are flagged only as
+         * `isDifferent`, and we have to reconcile them.
+         */
+
```
Comment (skinny85) [inline packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:101]: Same here (redundant empty line).

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts [hunk 3]
```diff
@@ -64,43 +69,94 @@ async function isLambdaFunctionCodeOnlyChange(
...
+        if (updatedProp !== undefined) {
```
Comment (skinny85) [inline packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:102]: What if `updatedProp` is `undefined`? Does this mean we need to remove all existing Tags?

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts [hunk 4]
```diff
@@ -64,43 +69,94 @@ async function isLambdaFunctionCodeOnlyChange(
...
+          foundDifference = foundDifference || updatedProp.isDifferent;
```
Comment (skinny85) [inline packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:103]: Hmm. What if we changed the above `if` statement to be `if (updatedProp?.isDifferent) {`? That would allow us to simplify this to just `foundDifference = true;`.

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts [hunk 5]
```diff
@@ -64,43 +69,94 @@ async function isLambdaFunctionCodeOnlyChange(
...
+  DELETE = -1,
+}
+
+interface LambdaFunctionTags {
+  readonly tagUpdates: { [tag : string] : string | TagDeletion };
+}
+
+interface LambdaFunctionChange {
+  code?: LambdaFunctionCode,
+  tags?: LambdaFunctionTags,
```
Comment (skinny85) [inline packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:154]: Are these actually optional? Feels like they should be required.

Oh, and also `readonly` 🙂.

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts [hunk 6]
```diff
@@ -110,10 +166,50 @@ class LambdaFunctionHotswapOperation implements HotswapOperation {
...
+        S3Key: resource.code.s3Key,
+      }).promise();
+    }
+
+    if (resource.tags !== undefined) {
+      // It's really unfortunate, but Lambda allows function _names_ for code
+      // updates, but we need the function ARN if we're managing tags. We also
+      // can't assume we'll get it from the preceding code update, since there
+      // might not _be_ a code update.
+      const tagsToDelete = Object.entries(resource.tags!.tagUpdates)
```
Comment (skinny85) [inline packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:185]: I believe you don't need the `!` here (TypeScript has flow-aware typing).

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts [hunk 7]
```diff
@@ -110,10 +166,50 @@ class LambdaFunctionHotswapOperation implements HotswapOperation {
...
+        .filter(([_key, val]) => val === TagDeletion.DELETE)
+        .map(([key, _val]) => key);
+
+      const tagsToSet: { [tag: string]: string } = {};
+      Object.entries(resource.tags!.tagUpdates)
```
Comment (skinny85) [inline packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:190]: Same here (redundant `!`).

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts (hunk 6 above)
Comment (skinny85) [inline packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:185]: Can you specify the type of `tagsToDelete` explicitly here, like you do for `tagsToSet`?

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts [hunk 8]
```diff
@@ -110,10 +166,50 @@ class LambdaFunctionHotswapOperation implements HotswapOperation {
...
-    }).promise();
+    const lambda = sdk.lambda();
+    const resource = this.lambdaFunctionResource.resource;
+
+    if (resource.code !== undefined) {
+      await lambda.updateFunctionCode({
+        FunctionName: this.lambdaFunctionResource.physicalName,
+        S3Bucket: resource.code.s3Bucket,
+        S3Key: resource.code.s3Key,
+      }).promise();
```
Comment (skinny85) [inline packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:177]: Any reason we're waiting here for the Lambda code to update first before handling the tags? Perhaps we can do these in parallel?

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts [hunk 9]
```diff
@@ -110,10 +166,50 @@ class LambdaFunctionHotswapOperation implements HotswapOperation {
...
+          TagKeys: tagsToDelete,
+        }).promise();
+      }
+
+      if (Object.keys(tagsToSet).length > 0) {
+        await lambda.tagResource({
+          Resource: configuration.FunctionArn!,
+          Tags: tagsToSet,
+        }).promise();
+      }
```
Comment (skinny85) [inline packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:212]: Same question here - can we do these in parallel? The set of tags to delete and to apply doesn't contain any common tags I'm 99% sure.

File: packages/cdk-assets/lib/private/shell.ts [hunk 10]
```diff
@@ -55,7 +55,7 @@ export async function shell(command: string[], options: ShellOptions = {}): Prom
         resolve(Buffer.concat(stdout).toString('utf-8'));
       } else {
         const out = Buffer.concat(stderr).toString('utf-8').trim();
-        reject(new ProcessFailed(code, `${renderCommandLine(command)} exited with error code ${code}: ${out}`));
+        reject(new ProcessFailed(code!, `${renderCommandLine(command)} exited with error code ${code}: ${out}`));
       }
     });
   });
```
Comment (skinny85) [inline packages/cdk-assets/lib/private/shell.ts:61]: Hmm... why these changes here?

File: packages/aws-cdk/test/assets.test.ts [hunk 11]
```diff
@@ -13,6 +13,8 @@ beforeEach(() => {
   assets = new AssetManifestBuilder();
 });
 
+const moduleExt = __filename.split('.').pop();
```
Comment (skinny85) [inline packages/aws-cdk/test/assets.test.ts:16]: Let's use [`path.extname()` instead](https://nodejs.org/api/path.html#pathextnamepath).

File: packages/aws-cdk/test/api/hotswap/hotswap-test-setup.ts [hunk 12]
```diff
@@ -81,9 +81,17 @@ export class HotswapMockSdkProvider {
     });
   }
 
-  public setUpdateFunctionCodeMock(mockUpdateLambdaCode: (input: lambda.UpdateFunctionCodeRequest) => lambda.FunctionConfiguration) {
+  public setUpdateFunctionCodeMock(
```
Comment (skinny85) [inline packages/aws-cdk/test/api/hotswap/hotswap-test-setup.ts:84]: The name of this method doesn't really fit anymore.

Let's just call this `stubLambda()` instead, and surface the same API `MockSdkProvider` does (it's nicer, because the object simulates named parameters).

File: packages/aws-cdk/test/api/hotswap/hotswap-deployments.test.ts [hunk 13]
```diff
@@ -3,15 +3,25 @@ import * as setup from './hotswap-test-setup';
...
+  mockGetFunctionConfiguration = jest.fn(({ FunctionName }) => {
+    return {
+      FunctionArn: `aws:::${FunctionName}`,
+    };
+  });
   mockUpdateMachineDefinition = jest.fn();
   mockGetEndpointSuffix = jest.fn(() => 'amazonaws.com');
-  hotswapMockSdkProvider.setUpdateFunctionCodeMock(mockUpdateLambdaCode);
+  hotswapMockSdkProvider.setUpdateFunctionCodeMock(mockUpdateLambdaCode, mockTagResource, mockUntagResource, mockGetFunctionConfiguration);
   hotswapMockSdkProvider.setUpdateStateMachineMock(mockUpdateMachineDefinition);
```
Comment (skinny85) [inline packages/aws-cdk/test/api/hotswap/hotswap-deployments.test.ts:25]: Since w're not referencing these variables (`mockTagResource`, `mockUntagResource`, `mockUntagResource`, `mockGetFunctionConfiguration`) anywhere, can we please have `setUpdateFunctionCodeMock()` just set them for us to default values (just simple Jest mocks are fine)?

File: packages/aws-cdk/test/api/hotswap/lambda-hotswap-deployments.test.ts [hunk 14]
```diff
@@ -80,6 +90,88 @@ test('calls the updateLambdaCode() API when it receives only a code difference i
...
+  });
+
+  expect(mockTagResource).toHaveBeenCalledWith({
+    Resource: 'aws:::my-function',
+    Tags: {
+      'to-be-changed': 'new-tag-value',
+      'to-be-added': 'added-tag-value',
+    },
+  });
+
```
Comment (skinny85) [inline packages/aws-cdk/test/api/hotswap/lambda-hotswap-deployments.test.ts:172]: No need for this empty line.

File: packages/aws-cdk/test/api/hotswap/lambda-hotswap-deployments.test.ts [hunk 15]
```diff
@@ -80,6 +90,88 @@ test('calls the updateLambdaCode() API when it receives only a code difference i
...
+});
```
Comment (skinny85) [inline packages/aws-cdk/test/api/hotswap/lambda-hotswap-deployments.test.ts:173]: Should we also assert that `updateFunctionCode` was _not_ called?

**Review by skinny85 — DISMISSED** [review]
Looks great @offbyone! A few comments, but nothing major.

**Comment by skinny85** [comment]
BTW, the build is failing with:

```
aws-cdk: FAIL test/api/exec.test.ts
aws-cdk:   â— Test suite failed to run
aws-cdk:     TypeError: Cannot read property 'DEFAULT' of undefined
aws-cdk:       14 | };
aws-cdk:       15 |
aws-cdk:     > 16 | export let logLevel = LogLevel.DEFAULT;
aws-cdk:          |                                ^
aws-cdk:       17 |
aws-cdk:       18 | export function setLogLevel(newLogLevel: LogLevel) {
aws-cdk:       19 |   logLevel = newLogLevel;
aws-cdk:       at Object.<anonymous> (lib/logging.ts:16:32)
aws-cdk:       at Object.<anonymous> (lib/api/cxapp/exec.ts:6:1)
```

So, you might have to revert your changes to the test setup.

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts [hunk 16]
```diff
@@ -110,10 +166,50 @@ class LambdaFunctionHotswapOperation implements HotswapOperation {
...
+      const tagsToSet: { [tag: string]: string } = {};
+      Object.entries(resource.tags!.tagUpdates)
+        .filter(([_key, val]) => val !== TagDeletion.DELETE)
+        .forEach(([tagName, tagValue]) => {
+          tagsToSet[tagName] = tagValue as string;
+        });
+
+      const configuration = await lambda.getFunctionConfiguration({
+        FunctionName: this.lambdaFunctionResource.physicalName,
+      }).promise();
```
Comment (skinny85) [inline packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:198]: Can we formulate the ARN ourselves, instead of making a call for it? We have the physical name from the previous part.

File: packages/aws-cdk/jest.config.js [hunk 17]
```diff
@@ -1,10 +1,22 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
 const baseConfig = require('@aws-cdk/cdk-build-tools/config/jest.config');
-module.exports = {
+module.exports = conf = {
```
[the remark this answers is not in the document]
Comment (offbyone) [reply packages/aws-cdk/jest.config.js:3]: I was originally modifying it after the assignment. I don't need to any more.

File: packages/aws-cdk/jest.config.js [hunk 18]
```diff
@@ -1,10 +1,22 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
```
[the remark this answers is not in the document]
Comment (offbyone) [reply packages/aws-cdk/jest.config.js:1]: I copied it from the jest docs. I honestly have no idea, and am happy to remove it.

File: packages/aws-cdk/lib/api/hotswap/common.ts [hunk 19]
```diff
@@ -54,7 +54,10 @@ export class HotswappableChangeCandidate {
    */
   public readonly propertyUpdates: { [key: string]: cfn_diff.PropertyDifference<any> };
 
-  public constructor(newValue: cfn_diff.Resource, propertyUpdates: { [key: string]: cfn_diff.PropertyDifference<any> }) {
+
+  public constructor(
+    newValue: cfn_diff.Resource, propertyUpdates: { [key: string]: cfn_diff.PropertyDifference<any> },
+  ) {
```
[the remark this answers is not in the document]
Comment (offbyone) [reply packages/aws-cdk/lib/api/hotswap/common.ts:60]: It is _shockingly_ hard to convince my editor _not_ to run the tidy rules from this package on code as I save it. I gave up after having to rewrite this diff a dozen times. These reformats are based off of the package level eslint, as far as I can tell. I _can_ undo them... maybe?

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts [hunk 20]
```diff
@@ -15,22 +16,25 @@ export async function isHotswappableLambdaFunctionChange(
...
-    // (for example, going from Code.fromAsset() to Code.fromInline())
+    // warn if the CDK asset metadata isn't present on the asset. We don't block
+    // hotswap here. This occurs when the assets are not built on top of the CDK
+    // native S3 and Image asset classes, which store asset metadata in them for the
+    // SAM CLI. We rely on the `isLambdaFunctionCodeOnlyChange` method to assure
+    // ourselves that this is not a structurally invalid change.
+
     if (!assetMetadataChanged(change)) {
-      return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+      warn('⚠️No cdk asset metadata is found in the change.');
```
[the remark this answers is not in the document]
Comment (offbyone) [reply packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:26]: Sure.

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts [hunk 21]
```diff
@@ -15,22 +16,25 @@ export async function isHotswappableLambdaFunctionChange(
...
     }
 
     const functionName = await establishResourcePhysicalName(logicalId, change.newValue.Properties?.FunctionName, evaluateCfnTemplate);
+
```
[the remark this answers is not in the document]
Comment (offbyone) [reply packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:30]: Vertical space is cheap, and I prefer it for readability.

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts (hunk 3 above)
Comment (offbyone) [reply packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:102]: No. If it's undefined, we don't modify the tags at all. We only explicitly remove/put tags that are identified in the diff.

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts (hunk 4 above)
Comment (offbyone) [reply packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:103]: I like it!

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts (hunk 5 above)
Comment (offbyone) [reply packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:154]: That actually brought my attention to an inconsistency above. Expect the next rev to tidy those up a bit.

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts (hunk 8 above)
Comment (offbyone) [reply packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:177]: I'm happy to, if you insist. I don't think it's a big deal -- tag updates are _fast_ -- but if you can walk me through what that code should look like I will (I don't fully grok how parallel code in JS works, sorry to say).

File: packages/cdk-assets/lib/private/shell.ts (hunk 10 above)
Comment (offbyone) [reply packages/cdk-assets/lib/private/shell.ts:61]: Incidental. IIRC I ran into those when trying to run the CDK using typescript instead of compiling it first.

File: packages/aws-cdk/test/api/hotswap/hotswap-deployments.test.ts (hunk 13 above)
Comment (offbyone) [reply packages/aws-cdk/test/api/hotswap/hotswap-deployments.test.ts:25]: The thing is, we _do_ reference them later in the tests. How would you rewrite the tests to do this? (see for example around lines 66-67)

Comment (skinny85) [reply packages/aws-cdk/test/api/hotswap/hotswap-deployments.test.ts:25]: Not sure what you mean. This is the entire diff in this file:
Screen Shot 2021-12-08 at 14 33 56

Comment (offbyone) [reply packages/aws-cdk/test/api/hotswap/hotswap-deployments.test.ts:25]: Oh, I see! Yeah, can do.

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts (hunk 8 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:172]: Sure. Basically, use `Promise.all()` instead of waiting for each promise before creating the new one.

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts (hunk 16 above)
Comment (skinny85) [reply packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:193]: Yes, we should. We do that already in `EvaluateCloudFormationTemplate`.

File: packages/aws-cdk/lib/api/hotswap/common.ts [hunk 22]
```diff
@@ -54,6 +54,7 @@ export class HotswappableChangeCandidate {
    */
   public readonly propertyUpdates: { [key: string]: cfn_diff.PropertyDifference<any> };
 
+
```
Comment (skinny85) [inline packages/aws-cdk/lib/api/hotswap/common.ts:57]: Let's revert this one:

```suggestion
```

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts [hunk 23]
```diff
@@ -15,22 +15,20 @@ export async function isHotswappableLambdaFunctionChange(
...
-    // (for example, going from Code.fromAsset() to Code.fromInline())
-    if (!assetMetadataChanged(change)) {
-      return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
-    }
-
     const functionName = await establishResourcePhysicalName(logicalId, change.newValue.Properties?.FunctionName, evaluateCfnTemplate);
+    const functionArn = await evaluateCfnTemplate.evaluateCfnExpression({
+      'Fn::Sub': 'arn:${AWS::Partition}:lambda:${AWS::Region}:${AWS::AccountId}:function:' + functionName,
+    });
+
```
Comment (skinny85) [inline packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:22]: Let's kill this empty line (I'm actually surprised the linter didn't flag this one):

```suggestion
```

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts [hunk 24]
```diff
@@ -63,44 +62,102 @@ async function isLambdaFunctionCodeOnlyChange(
    * even if only one of them was actually changed,
    * which means we don't need the "old" values at all, and we can safely initialize these with just `''`.
    */
-  let s3Bucket = '', s3Key = '';
-  let foundCodeDifference = false;
   // Make sure only the code in the Lambda function changed
   const propertyUpdates = change.propertyUpdates;
+  let code: LambdaFunctionCode | void = undefined;
+  let tags: LambdaFunctionTags | void = undefined;
```
Comment (skinny85) [inline packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:68]:
```suggestion
  let code: LambdaFunctionCode | undefined = undefined;
  let tags: LambdaFunctionTags | undefined = undefined;
```

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts [hunk 25]
```diff
@@ -63,44 +62,102 @@ async function isLambdaFunctionCodeOnlyChange(
...
+
+          updatedProp.oldValue.forEach((tag: CfnDiffTagValue) => {
+            if (tagUpdates[tag.Key] === undefined) {
+              tagUpdates[tag.Key] = TagDeletion.DELETE;
+            }
+          });
+
+          tags = {
+            tagUpdates,
+          };
```
Comment (skinny85) [inline packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:118]:
```suggestion
          tags = { tagUpdates };
```

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts [hunk 26]
```diff
@@ -63,44 +62,102 @@ async function isLambdaFunctionCodeOnlyChange(
...
+  DELETE = -1,
+}
+
+interface LambdaFunctionTags {
+  readonly tagUpdates: { [tag : string] : string | TagDeletion };
+}
+
+interface LambdaFunctionChange {
+  readonly code: LambdaFunctionCode | void,
+  readonly tags: LambdaFunctionTags | void,
```
Comment (skinny85) [inline packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:154]:
```suggestion
  readonly code?: LambdaFunctionCode;
  readonly tags?: LambdaFunctionTags;
```

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts [hunk 27]
```diff
@@ -110,10 +167,51 @@ class LambdaFunctionHotswapOperation implements HotswapOperation {
...
+        S3Bucket: resource.code.s3Bucket,
+        S3Key: resource.code.s3Key,
+      }).promise());
+    }
+
+    if (resource.tags !== undefined) {
+      // It's really unfortunate, but Lambda allows function _names_ for code
+      // updates, but we need the function ARN if we're managing tags. We also
+      // can't assume we'll get it from the preceding code update, since there
+      // might not _be_ a code update.
```
Comment (skinny85) [inline packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:186]: No need for this comment anymore I think:

```suggestion
```

File: packages/aws-cdk/lib/api/hotswap/lambda-functions.ts [hunk 28]
```diff
@@ -110,10 +167,51 @@ class LambdaFunctionHotswapOperation implements HotswapOperation {
...
+      if (Object.keys(tagsToSet).length > 0) {
+        operations.push(lambda.tagResource({
+          Resource: this.lambdaFunctionResource.functionArn,
+          Tags: tagsToSet,
+        }).promise());
+      }
+    }
+
+    // run all of our updates in parallel
+    await Promise.all(operations);
```
Comment (skinny85) [inline packages/aws-cdk/lib/api/hotswap/lambda-functions.ts:215]:
```suggestion
    return Promise.all(operations);
```

File: packages/aws-cdk/lib/logging.ts [hunk 29]
```diff
@@ -13,6 +13,15 @@ const logger = (stream: Writable, styles?: StyleFn[]) => (fmt: string, ...args:
...
 };
 
+export const enum LogLevel {
+  /** Not verbose at all */
+  DEFAULT = 0,
+  /** Pretty verbose */
+  DEBUG = 1,
+  /** Extremely verbose */
+  TRACE = 2
+}
```
Comment (skinny85) [inline packages/aws-cdk/lib/logging.ts:23]: Let's revert this move too (unless there's a good reason for it...?).

**Review by skinny85 — COMMENTED** [review]
Looks great overall! Some minor comments, mainly around formatting.

@rix0rrr do you mind taking a quick look at the test setup changes? Thanks!

**Review by skinny85 — APPROVED** [review]

**Review by skinny85 — APPROVED** [review]
